microprofile/WFLY-12313_MicroProfile_OpenAPI_1.1.adoc

```diff
@@ -0,0 +1,111 @@
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-12313
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:fburzigo@redhat.com[Fabio Burzigotti]
+
+=== Testing By
+
+[ ] Engineering
+
+[X] QE
+
+=== Affected Projects or Components
+
+The SmallRye implementation of the MicroProfile OpenAPI specification is already integrated by the following projects:
+
+* https://thorntail.io/[Thorntail] via the https://github.com/thorntail/thorntail/tree/master/fractions/microprofile/microprofile-openapi[microprofile-openapi fraction].
+* https://quarkus.io/[Quarkus] via the https://github.com/quarkusio/quarkus/tree/master/extensions/smallrye-openapi[smallrye-openapi extension].
+
+These integrations will heavily inform, though not dictate, the corresponding integration into WildFly.
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* Create an MP OpenAPI galleon layer
+** Should depend on MP Config and JAX-RS (and their dependencies)
+* The microprofile-openapi spec module must be exposed to deployments
+* The `/openapi` endpoint shall be registered relative to the root of the virtual host associated with the deployment.
+** This places a superficial constraint of a single application per host (see subsequent section re: multiple deployments)
+** In the event a web deployment uses a given host for which another deployment has already registered an `/openapi` endpoint, we will:
+... Log a warning.
+... Skip OpenAPI model generation for that deployment.
+* All WildFly/SmallRye specific configuration will be obtained via MP Config using the properties prefixed with `mp.openapi.extensions`.
+** Consequently, the microprofile-openapi subsystem should have an empty management model.
+* Support YAML and JSON for both input (static/pre-generated OpenAPI document) and output formats.
+* MicroProfile OpenAPI TCK must pass
+
+=== Nice-to-Have Requirements
+
+* Optional features of the specification:
+** Multiple deployments:
+*** If a container allows multiple deployments, the specification expects the `/openapi` endpoint to return the logical union of the documented services for all applications in the runtime. See: https://download.eclipse.org/microprofile/microprofile-open-api-1.1.2/microprofile-openapi-spec.html#_multiple_applications
+**** Initially, we will limit multiple deployment support to a single deployment per host.
+**** Merging (and more interestingly, unmerging on undeploy) OpenAPI models for all applications deployed to a given host.
+***** Consider for a future enhancement
+**** Until full multiple deployment support is available, users might want/need the ability to disable (i.e. opt out of) OpenAPI documentation of a given web deployment, e.g. to support deployment of multiple WARs associated with the same virtual host, where only 1 is a JAX-RS application.
+***** Consider for a future enhancement
```

Comment:
I think this should be part of this effort, users need to have a way to specify preferred deployment for `/openapi` exposure.

Users can have *.ear with multiple *.war archives inside or just simply deploying multiple *.war archives.

Reply:
My point is that this shouldn't be marked as "Consider for a future enhancement"

Reply:
@rsvoboda I agree and given previous comments regarding this topic i think that this should be covered by a new hard requirement ( "enable/disable `/openapi` endpoint registration on a per-application basis via MP config). @pferraro WDYT?